### PR TITLE
boosting friends keep by recency

### PR DESCRIPTION
**Pull request by ymatsuda**
- a friend's keep is boosted according to the oldest timestamp
  - ex) if three users keep pageX, we take the oldest createdAt of their keeps as the timestamp.
- we add one most recent keep in the result if, in after regular processing, 
  - we have a space in the hit list, and
  - this keep is not in the hit list yet
- added "isNew" flag to PersonalSearchResult
  - the flag is true is a friend kept the page within a day (we will revisit this later)
  - this is for the frontend to use for rendering hits

### CODE REVIEW

File: shoebox/app/com/keepit/serializer/PersonalSearchResultSerializer.scala [hunk 1]
```diff
@@ -16,7 +16,8 @@ class PersonalSearchResultSerializer extends Writes[PersonalSearchResult] with L
         "users" -> Json.toJson(res.users),
         "score" -> JsNumber(res.score),
         "isMyBookmark" -> JsBoolean(res.isMyBookmark),
-        "isPrivate" -> JsBoolean(res.isPrivate)
+        "isPrivate" -> JsBoolean(res.isPrivate),
+        "isNew" -> JsBoolean(res.isNew)
```
Comment (2is10) [inline shoebox/app/com/keepit/serializer/PersonalSearchResultSerializer.scala]: Are "new" results going to have a different visual treatment? If not, I think I'd lean towards keeping "newness" invisible to search consumers, to keep them as simple as possible.

File: shoebox/app/com/keepit/controllers/ext/ExtSearchController.scala [hunk 2]
```diff
@@ -147,7 +147,7 @@ class ExtSearchController @Inject() (
         val users = res.hits.map(_.users).flatten.distinct.map(u => u -> basicUserRepo.load(u)).toMap
         log.info(s"search-personal-a: ${currentDateTime.getMillis()-t0}")
         val t1 = currentDateTime.getMillis()
-        val h = res.hits.map(toPersonalSearchResult(userId, users, _))
+        val h = res.hits.zip(res.scorings).map{ case (hit, scoring) => toPersonalSearchResult(userId, users, hit, scoring) }
```
Comment (gmethvin) [inline shoebox/app/com/keepit/controllers/ext/ExtSearchController.scala]: You could use some fancy scala shorthand if you like:

``` scala
(res.hits zip res.scorings) map { toPersonalSearchResult(userId, users, _, _) }
```

File: shoebox/app/com/keepit/controllers/ext/ExtSearchController.scala [hunk 3]
```diff
@@ -166,9 +166,18 @@ class ExtSearchController @Inject() (
...
-      bookmark.map(_.isPrivate).getOrElse(false), users, res.score)
+
+    // we mark the friend keep "isNew" if recencyScore > 0.5 which means the oldest create time on the hits is
+    // within the halfDecay period. recencyScore is always zero for others' keep.
+    val isNew = (!res.isMyBookmark && scoring.recencyScore > 0.5f)
+
+    PersonalSearchResult(toPersonalSearchHit(uri, bookmark),
+                         res.bookmarkCount,
+                         res.isMyBookmark,
+                         bookmark.map(_.isPrivate).getOrElse(false),
```
Comment (gmethvin) [inline shoebox/app/com/keepit/controllers/ext/ExtSearchController.scala]: `bookmark.exists(_.isPrivate)`

File: shoebox/app/com/keepit/serializer/PersonalSearchResultSerializer.scala (hunk 1 above)
Comment (gmethvin) [inline shoebox/app/com/keepit/serializer/PersonalSearchResultSerializer.scala]: agreed; we can always add things to the public api later if we need it

File: shoebox/app/com/keepit/controllers/ext/ExtSearchController.scala (hunk 2 above)
Comment (ymatsuda) [inline shoebox/app/com/keepit/controllers/ext/ExtSearchController.scala]: cool. I will take it.

File: shoebox/app/com/keepit/serializer/PersonalSearchResultSerializer.scala (hunk 1 above)
Comment (ymatsuda) [inline shoebox/app/com/keepit/serializer/PersonalSearchResultSerializer.scala]: ok. I will revert the change.

File: shoebox/app/com/keepit/controllers/ext/ExtSearchController.scala (hunk 2 above)
Comment (ymatsuda) [inline shoebox/app/com/keepit/controllers/ext/ExtSearchController.scala]: This doesn't compile. :(
wrong number of parameters; expected = 1

File: shoebox/app/com/keepit/controllers/ext/ExtSearchController.scala (hunk 3 above)
Comment (ymatsuda) [inline shoebox/app/com/keepit/controllers/ext/ExtSearchController.scala]: nice!

**Pull request merged**